### ISSUE TITLE
BatchNorm in test setting.

### ISSUE THREAD

**Issue by Lan1991Xu** [Britefury/self-ensemble-visual-domain-adapt-photo#3]
Hi Thanks for your sharing code. I have noticed you have used different batch norm parameters in two domain. How about in the testing. I only find these code part related to batch norm setting in model_architecture.py, could you explain in detail. Thanks

**Comment by Britefury** [comment]
The reason you cannot find a specific implementation is that there is none :)
It occurs as a side effect of how the batches are processed.

If data from both datasets is processed in a single batch:

```py3
combined_X = np.concatenate([source_dataset_X, target_dataset_X], axis=0)
combined_y = np.concatenate([source_dataset_y, target_dataset_y], axis=0)
combined_prediction = net(combined_X)
loss = criterion(combined_prediction, combined_y)
```

The batch norm stats (mean, std) will be computed using samples from both source and target

If processed separately:

```py3
source_prediction = net(source_dataset_X)
target_prediction = net(target_dataset_X)
source_loss = criterion(source_prediction, source_dataset_y)
target_loss = criterion(target_prediction, target_dataset_y)
```

Then batch norm will be applied to the source and target batches separately, using a different mean and std for each.

**Comment by Lan1991Xu** [comment]
Thanks for you explain. I think I just misunderstand in the test stage.  Actually you used the mean and variance in the training, when apply BN in batch norm. So in the test stage, there are not separated “the mean and variance”.  But then I come with an issue, why we need class DomainAdaptModule (nn.Module)?
If we do not want to keep the statistic information for source and target separately. Can we directly use the default batch norm in pytorch. Is there any difference here? Thanks again.

**Comment by lianqing11** [comment]
@Britefury I have another question about the BatchNorm. Because the teacher model only collect target domain batch normalization information(mean and var), it will be better than collect both the source and the target domain. You can check it from https://arxiv.org/abs/1603.04779 this paper.

I am wondering that if the prediction difference on student and teacher model maybe also not only come from different augmentation but also come from different bn information.
